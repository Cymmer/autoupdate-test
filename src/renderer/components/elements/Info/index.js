import React from 'react';
import PropTypes from 'prop-types';
import GLOBALS from 'codechum-app-globals';
import styles from './styles.module.scss';

import Text from '../Text';
import textTypes from '../Text/constants/textTypes';

import infoSizes from './constants/infoSizes';
import Tooltip from '../Tooltip';
import { tooltipPlacement } from '../Tooltip/constants';
import Icon from '../Icon';

const Info = ({ label, children, className, size, id, tooltip }) => (
  <div className={className} id={id}>
    <Text
      className={styles.Info_label}
      colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['500']}
      type={size === infoSizes.MD ? textTypes.BODY.MD : textTypes.BODY.SM}
    >
      {label}
      {tooltip !== null && (
        <Tooltip
          className={styles.Info_label_tooltip}
          content={tooltip}
          placement={tooltipPlacement.RIGHT}
        >
          <Icon icon="info" />
        </Tooltip>
      )}
    </Text>
    <Text
      className={styles.Info_text}
      type={size === infoSizes.MD ? textTypes.BODY.MD : textTypes.BODY.SM}
    >
      {children}
    </Text>
  </div>
);

Info.defaultProps = {
  className: null,
  id: null,
  size: infoSizes.MD,
  tooltip: null,
};

Info.propTypes = {
  tooltip: PropTypes.string,
  label: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  id: PropTypes.string,
  size: PropTypes.oneOf([infoSizes.MD, infoSizes.SM]),
};

export default Info;
