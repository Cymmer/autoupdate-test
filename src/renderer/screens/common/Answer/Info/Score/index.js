import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import {
  determinePercentageColorName,
  determinePercentageColorClass,
} from 'codechum-app-utils';
import GLOBALS from 'codechum-app-globals';
import styles from './styles.module.scss';

import {
  Card,
  Text,
  Span,
  Tooltip,
  Icon,
} from '../../../../../components/elements';
import {
  textTypes,
  tooltipPlacement,
} from '../../../../../components/elements/constants';
import { useInterval } from '../../../../../hooks';

const COOLDOWN_SECONDS = 25;

const Score = ({ answer, isMustBePerfectShown, task }) => {
  const [pulsatingCountdown, setPulsatingCountdown] = useState(0);
  const [initialSubmissions, setInitialSubmissions] = useState(null);

  const {
    score,
    question: { max_score: maxScore },
    submissions,
  } = answer || {};

  useEffect(() => {
    if (!initialSubmissions) {
      // we do this to prevent the tooltip from showing on load
      setInitialSubmissions(submissions);
    } else if (submissions != null && isMustBePerfectShown) {
      setPulsatingCountdown(COOLDOWN_SECONDS);
    }
  }, [submissions]);

  useInterval(() => {
    if (pulsatingCountdown > 0) {
      setPulsatingCountdown(pulsatingCountdown - 1);
    }
  }, 1000);

  return (
    <Card className={styles.Score}>
      <div className={styles.Score_info}>
        <Text className={styles.Score_text} type={textTypes.DATA.LG}>
          Score:{' '}
          {task?.is_delayed_grading ? (
            <Tooltip
              content="Your teacher has chosen to check your work manually. Your score will be available once your teacher is finished grading your work."
              placement={tooltipPlacement.RIGHT}
            >
              <Span colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['500']}>
                -/{maxScore}
              </Span>
            </Tooltip>
          ) : (
            <Span
              colorClass={determinePercentageColorClass(
                parseInt(score),
                maxScore
              )}
            >
              {parseInt(score)}/{maxScore}
            </Span>
          )}
        </Text>
        {pulsatingCountdown > 0 && (
          <Tooltip
            className={styles.Score_tooltip}
            content={
              <span className={styles.Score_tooltip_content}>
                All test cases must be correct for this problem in order for
                your score to be recorded
              </span>
            }
            placement="right"
            theme="danger"
          >
            <Icon isPulsating className={styles.Score_dangerIcon} icon="info" />
          </Tooltip>
        )}
      </div>
      <progress
        className={
          styles[
            `Score_progress___${determinePercentageColorName(
              parseInt(score),
              maxScore
            )}`
          ]
        }
        max={maxScore}
        value={parseInt(score)}
      />
    </Card>
  );
};

Score.defaultProps = {
  answer: null,
  isMustBePerfectShown: false,
};

Score.propTypes = {
  answer: PropTypes.object,
  task: PropTypes.object,

  // if true, the must be perfect notice will be shown
  isMustBePerfectShown: PropTypes.bool,
};

export default Score;
