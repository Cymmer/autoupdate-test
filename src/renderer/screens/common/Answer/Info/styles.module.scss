@import '../../../../scss/mixins';

.Info {
  @include flex-between;
  flex-direction: column;
  height: 100%;
  padding: 8px;
  min-width: 300px;
  max-height: calc(100vh - 60px);

  @include breakpoint(sm) {
    max-height: none;
  }

  &_content {
    height: 100%;
    width: 100%;
  }

  &_actions {
    width: 100%;
    margin: 8px 0 0 0;

    @include breakpoint(sm) {
      padding: 0 0 80px 0;
    }
  }
}
