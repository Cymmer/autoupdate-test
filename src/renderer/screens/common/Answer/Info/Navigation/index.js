import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';
import GLOBALS from 'codechum-app-globals';

import { Card, Text } from 'components/elements';
import { textTypes } from 'components/elements/constants';
import styles from './styles.module.scss';

const Navigation = ({
  menuButton,
  leftButton,
  navigationText,
  rightButton,
  submitButton,
  className,
}) => (
  <Card className={cn(styles.Navigation, className)}>
    {menuButton}
    <div className={styles.Navigation_nav}>
      {leftButton}

      {navigationText && (
        <Text
          className={styles.Navigation_nav_text}
          colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
          type={textTypes.BODY.MD}
        >
          {navigationText}
        </Text>
      )}

      {rightButton}

      {submitButton}
    </div>
  </Card>
);

Navigation.defaultProps = {
  menuButton: null,

  // left button, right button, and navigation text
  // may be null if the task info is toggled but
  // if one of them is null, THE REST MUST ALSO
  // be null
  leftButton: null,
  navigationText: null,
  rightButton: null,

  submitButton: null,
  className: null,
};

Navigation.propTypes = {
  // the button that'll toggle the menu (switches the question info
  // to task info and vice versa)
  menuButton: PropTypes.element,

  // the button that'll navigate to the previous problem
  leftButton: PropTypes.element,

  // the button that'll navigate to the next problem
  rightButton: PropTypes.element,

  // the text in between the button
  navigationText: PropTypes.string,

  // the button that'll toggle the Are You Sure modal, in preparetion
  // for the submission of the whole Result of the student
  submitButton: PropTypes.element,

  className: PropTypes.string,
};

export default Navigation;
