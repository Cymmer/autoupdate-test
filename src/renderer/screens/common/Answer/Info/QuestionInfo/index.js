import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import GLOBALS from 'codechum-app-globals';

import { ProblemInputList, ProblemInput } from 'components';
import { Text, Code, Card, Span, ButtonLink } from 'components/elements';
import { buttonTypes, textTypes } from 'components/elements/constants';
import styles from './styles.module.scss';

const getActualCreatorName = (creator) => {
  if (creator.first_name) {
    return `${creator.first_name} ${creator.last_name}`;
  }

  return 'CodeChum';
};

const QuestionInfo = ({
  answer,
  number,
  showProblemSolution,
  descriptionRef,
  inputRef,
  outputRef,
  descriptionActiveOnboarding,
  inputActiveOnboarding,
  outputActiveOnboarding,
  isInteractive,
}) => {
  const { question } = answer || {};

  const {
    problem: {
      creator,
      name: problemName,
      description,
      inputs,
      output_format: outputFormat,
      sample_output: sampleOutput,
      creator_name_override: creatorNameOverride,
      problem_solution: problemSolution,
    },
    question_additional_info: questionAdditionalInfo,
  } = question || {};

  return (
    <Card className={styles.QuestionInfo}>
      <div
        ref={descriptionRef}
        className={cn({
          [styles.QuestionInfo___active_onboarding]:
            descriptionActiveOnboarding,
        })}
      >
        <div className={styles.QuestionInfo_head}>
          <div>
            <Text type={textTypes.HEADING.XS}>
              <Span colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}>
                {number}.
              </Span>{' '}
              {problemName}
            </Text>
            <Text
              className={styles.QuestionInfo_creator}
              colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
              type={textTypes.BODY.SM}
            >
              by {creatorNameOverride || getActualCreatorName(creator)}
            </Text>
          </div>
          {showProblemSolution && problemSolution && (
            <ButtonLink
              data-test="viewSolutionLink"
              icon="launch"
              to={problemSolution}
              type={buttonTypes.TEXT.BLUE}
            >
              View Solution
            </ButtonLink>
          )}
        </div>

        <div className={styles.QuestionInfo_description}>
          <div dangerouslySetInnerHTML={{ __html: description }} />
          {questionAdditionalInfo?.instruction != null && (
            <div className={styles.QuestionInfo_instructions}>
              <Text
                className={styles.QuestionInfo_label}
                colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
                type={textTypes.BODY.SM}
              >
                Instructions
              </Text>
              <div
                dangerouslySetInnerHTML={{
                  __html: questionAdditionalInfo.instruction,
                }}
              />
            </div>
          )}
        </div>
      </div>

      {inputs?.length > 0 && isInteractive && (
        <div
          ref={inputRef}
          className={cn(styles.QuestionInfo_format, {
            [styles.QuestionInfo___active_onboarding]: inputActiveOnboarding,
          })}
          data-test="interactiveInput"
        >
          <Text
            colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
            type={textTypes.BODY.SM}
          >
            Input
          </Text>
          <ProblemInputList>
            {inputs.map((input, index) => (
              <ProblemInput
                key={`${input.input_name}-${index}`}
                data-test={`problemInput-${input.input_name}`}
                input={input}
                number={index + 1}
              />
            ))}
          </ProblemInputList>
        </div>
      )}

      {inputs?.length > 0 && !isInteractive && inputs[0]?.input_sample && (
        <div
          ref={inputRef}
          className={cn(styles.QuestionInfo_format, {
            [styles.QuestionInfo___active_onboarding]: inputActiveOnboarding,
          })}
          data-test="nonInteractiveInput"
        >
          <div>
            <Text
              className={styles.QuestionInfo_label}
              colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
              type={textTypes.BODY.SM}
            >
              Input
            </Text>
            <Text
              className={styles.QuestionInfo_inputOutputFormat}
              type={textTypes.BODY.SM}
            >
              {inputs[0]?.input_description}
            </Text>
          </div>
          <Code>
            <div
              dangerouslySetInnerHTML={{
                __html: inputs[0]?.input_sample
                  .split(' ')
                  .join('<span class="midDot">·</span>'),
              }}
            />
          </Code>
        </div>
      )}

      <div
        ref={outputRef}
        className={cn(styles.QuestionInfo_format, {
          [styles.QuestionInfo___active_onboarding]: outputActiveOnboarding,
        })}
      >
        <div>
          <Text
            className={styles.QuestionInfo_label}
            colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
            type={textTypes.BODY.SM}
          >
            Output
          </Text>
          <Text
            className={styles.QuestionInfo_inputOutputFormat}
            type={textTypes.BODY.SM}
          >
            {outputFormat}
          </Text>
        </div>
        {sampleOutput ? (
          <Code>
            <div
              dangerouslySetInnerHTML={{
                __html: sampleOutput
                  .split(' ')
                  .join('<span class="midDot">·</span>'),
              }}
            />
          </Code>
        ) : (
          <Text type={textTypes.BODY.SM}>No Output</Text>
        )}
      </div>
    </Card>
  );
};

QuestionInfo.defaultProps = {
  showProblemSolution: false,
  isInteractive: false,
};

QuestionInfo.propTypes = {
  answer: PropTypes.shape({
    question: PropTypes.shape({
      problem: PropTypes.shape({
        creator: PropTypes.shape({
          first_name: PropTypes.string,
          last_name: PropTypes.string,
        }),
        name: PropTypes.string,
        description: PropTypes.string,
        input_format: PropTypes.string,
        sample_input: PropTypes.string,
        output_format: PropTypes.string,
        sample_output: PropTypes.string,
        creator_name_override: PropTypes.string,
        problem_solution: PropTypes.string,
      }),
      question_additional_info: PropTypes.shape({
        instruction: PropTypes.string,
      }),
    }),
  }).isRequired,
  number: PropTypes.number.isRequired,

  showProblemSolution: PropTypes.bool,

  // refs
  descriptionRef: PropTypes.object.isRequired,
  inputRef: PropTypes.object.isRequired,
  outputRef: PropTypes.object.isRequired,

  // onboarding active status
  descriptionActiveOnboarding: PropTypes.bool,
  inputActiveOnboarding: PropTypes.bool,
  outputActiveOnboarding: PropTypes.bool,

  isInteractive: PropTypes.bool,
};

export default QuestionInfo;
