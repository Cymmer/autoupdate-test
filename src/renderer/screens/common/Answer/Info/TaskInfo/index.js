import React from 'react';
import PropTypes from 'prop-types';

import { Card, Text } from 'components/elements';
import { textTypes } from 'components/elements/constants';
import styles from './styles.module.scss';

import ProblemLink from './ProblemLink';

const TaskInfo = ({ task: { name: taskName }, problemLinks }) => (
  <div className={styles.TaskInfo}>
    <Card className={styles.TaskInfo_card}>
      <Text type={textTypes.HEADING.XS}>{taskName}</Text>
    </Card>
    <div className={styles.TaskInfo_problems}>{problemLinks}</div>
  </div>
);

TaskInfo.ProblemLink = ProblemLink;

TaskInfo.propTypes = {
  task: PropTypes.object.isRequired,
  problemLinks: PropTypes.arrayOf(PropTypes.element).isRequired,
};

export default TaskInfo;
