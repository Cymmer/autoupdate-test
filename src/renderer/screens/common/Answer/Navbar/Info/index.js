import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import { Text, Span, Icon, Tooltip } from 'components/elements';
import { textTypes } from 'components/elements/constants';
import { STUDENT_ROUTES } from 'screen-wrappers/Student/constants';
import GLOBALS from 'codechum-app-globals';
import styles from '../styles.module.scss';

const Info = ({ label, colorClass, value, isLocked }) => (
  <Text
    className={styles.Navbar_info_text}
    colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['500']}
    type={textTypes.BODY.SM}
  >
    <Span
      className={cn(styles.Navbar_info_text_label, {
        [styles.Navbar_info_text_label___locked]: isLocked,
      })}
    >
      {label}
      {': '}
    </Span>
    {!isLocked ? (
      <Span colorClass={colorClass} type={textTypes.DATA.SM}>
        {' '}
        {value}
      </Span>
    ) : (
      <Tooltip content={GLOBALS.MESSAGES.PREMIUM_FEATURE_LOCK_MESSAGE}>
        <a
          className={styles.Navbar_info_text_lockedIcon_link}
          href={STUDENT_ROUTES.SHOP}
          rel="noreferrer"
          target="_blank"
        >
          <Icon className={styles.Navbar_info_text_lockedIcon} icon="lock" />
        </a>
      </Tooltip>
    )}
  </Text>
);

Info.defaultProps = {
  isLocked: false,
};

Info.propTypes = {
  label: PropTypes.string.isRequired,
  colorClass: PropTypes.oneOf([
    GLOBALS.COLOR_CLASSES.RED['300'],
    GLOBALS.COLOR_CLASSES.BLUE['300'],
    GLOBALS.COLOR_CLASSES.GREEN['300'],
    GLOBALS.COLOR_CLASSES.NEUTRAL['500'],
  ]).isRequired,
  value: PropTypes.string.isRequired,
  isLocked: PropTypes.bool,
};

export default Info;
