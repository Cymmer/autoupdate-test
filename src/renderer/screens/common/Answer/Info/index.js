import React from 'react';
import PropTypes from 'prop-types';
import styles from './styles.module.scss';

import QuestionInfo from './QuestionInfo';
import TaskInfo from './TaskInfo';
import Score from './Score';
import Navigation from './Navigation';

const Info = ({ taskInfo, score, questionInfo, navigation }) => (
  <div className={styles.Info}>
    <div className={styles.Info_content}>
      {questionInfo}
      {taskInfo}
    </div>
    <div className={styles.Info_actions}>
      {score}

      {navigation}
    </div>
  </div>
);

Info.QuestionInfo = QuestionInfo;
Info.Score = Score;
Info.Navigation = Navigation;
Info.TaskInfo = TaskInfo;

Info.defaultProps = {
  taskInfo: null,
  score: null,
  questionInfo: null,
  navigation: null,
};

Info.propTypes = {
  // JSX element of type Info.TaskInfo
  taskInfo: PropTypes.node,

  // JSX element of type Info.Score
  score: PropTypes.node,

  // JSX element of type Info.QuestionInfo
  questionInfo: PropTypes.node,

  // JSX element of type Info.Navigation
  navigation: PropTypes.node,
};

export default Info;
