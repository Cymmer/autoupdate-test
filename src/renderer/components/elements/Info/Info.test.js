import React from 'react';
import { shallow } from 'enzyme';
import Info from './index';

describe('RENDERING - Info', () => {
  it('should render with default props', () => {
    const component = shallow(
      <Info label="Test">
        <p>Test</p>
      </Info>
    );
    expect(component).toMatchSnapshot();
  });
});
