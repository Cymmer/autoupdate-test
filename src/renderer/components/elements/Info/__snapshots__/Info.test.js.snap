// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDERING - Info should render with default props 1`] = `
<div
  className={null}
  id={null}
>
  <Text
    className="Info_label"
    colorClass="n___500"
    element={null}
    id={null}
    type="body___md"
  >
    Test
  </Text>
  <Text
    className="Info_text"
    colorClass="n___700"
    element={null}
    id={null}
    type="body___md"
  >
    <p>
      Test
    </p>
  </Text>
</div>
`;
